File: EMVS/SimEMVSS.R
if ( file.exists("~/project/EMSHS/EMVS/EMVS_Rcode_original.R") )
  source("~/project/EMSHS/EMVS/EMVS_Rcode_original.R")
if ( file.exists("EMVS/EMVS_Rcode_original.R") )
  source("EMVS/EMVS_Rcode_original.R")
#library(EMVS)

SimEMVSS <- function(r,v0,v1,eta=0,datapath,batch=0)
{
  D1 = length(v0)
  D2 = length(eta)

  FNrate = array(0,c(D1,D2,r))
  FPrate = array(0,c(D1,D2,r))
  MSTE = array(0,c(D1,D2,r))
  MSPE = array(0,c(D1,D2,r))
  time = array(0,c(D1,D2,r))
  
  for ( i in 1:r )
  {
    print(i)
    load(sprintf("%s/data%03d",datapath,batch+i))
    if ( data$p <= 10000 )
    {
      G = diag(0,data$p)
      G[data$E[,1]+(data$E[,2]-1)*data$p] = 1
    }

    for ( d2 in 1:D2 )
    {
      if ( eta[d2]==0 )
        time[,d2,i] = system.time(fit <- EMVS(data$y,data$X,v0=v0,v1=v1,type="betabinomial",sigma_init=1,epsilon=1e-3,a=1,b=1))[1]
      else
        time[,d2,i] = system.time(fit <- EMVS(data$y,data$X,v0=v0,v1=v1,type="MRF",mu=0,Sigma=G*eta[d2],sigma_init=1,epsilon=1e-3,v1_g=v1))[1]

      beta = t(fit$betas*(fit$p>0.5))

      FNrate[,d2,i] = apply(fit$p[,1:data$q]<0.5,1,mean)
      FPrate[,d2,i] = apply(fit$p[,(data$q+1):data$p]>0.5,1,mean)
    
      yhattune = data$Xtune %*% beta
      MSTE[,d2,i] = apply((data$ytune-yhattune)^2,2,mean)
      yhattest = data$Xtest %*% beta
      MSPE[,d2,i] = apply((data$ytest-yhattest)^2,2,mean)
    }
  }
  
  list(r=r,batch=batch,v0=v0,v1=v1,eta=eta,FNrate=FNrate,FPrate=FPrate,MSTE=MSTE,MSPE=MSPE,time=time)
}

SimEMVSS_old <- function(r,v0,v1,eta=0,datapath,batch=0)
{
  D1 = length(v0)
  D2 = length(eta)

  FNrate = array(0,c(D1,D2,r))
  FPrate = array(0,c(D1,D2,r))
  MSTE = array(0,c(D1,D2,r))
  MSPE = array(0,c(D1,D2,r))
  time = array(0,c(D1,D2,r))
  
  for ( i in 1:r )
  {
    print(i)
    load(sprintf("%s/data%03d",datapath,batch+i))
    G = diag(0,data$p)

    for ( d2 in 1:D2 )
    {
      G[data$E[,1]+(data$E[,2]-1)*data$p] = eta[d2]
      for ( d1 in 1:D1 )
      {
        if ( eta[d2]==0 )
          time[d1,d2,i] = system.time(fit <- EMVS(data$y,data$X,v0=v0[d1],v1=v1,type="betabinomial",sigma_init=1,epsilon=1e-3,a=1,b=1))[1]
        else
          time[d1,d2,i] = system.time(fit <- EMVS(data$y,data$X,v0=v0[d1],v1=v1,type="MRF",mu=0,Sigma=G,sigma_init=1,epsilon=1e-3,v1_g=v1))[1]

        beta = as.vector(fit$betas*(fit$p>0.5))

        FNrate[d1,d2,i] = mean(fit$p[1,1:data$q]<0.5)
        FPrate[d1,d2,i] = mean(fit$p[1,(data$q+1):data$p]>0.5)
    
        yhattune = data$Xtune %*% beta
        MSTE[d1,d2,i] = mean((data$ytune-yhattune)^2)
        yhattest = data$Xtest %*% beta
        MSPE[d1,d2,i] = mean((data$ytest-yhattest)^2)
      }
    }
  }
  
  list(r=r,batch=batch,v0=v0,v1=v1,eta=eta,FNrate=FNrate,FPrate=FPrate,MSTE=MSTE,MSPE=MSPE,time=time)
}

File: EMVS/SimEMVSS.c
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#define HPC 0
#define LPC 1
#define Emory 2

int main()
{
	char fname[100];
	char Rname[100];
	char line[1024];
	char method[20];
	char acronym[20];
	char master[100];
	char home[100];
	char script[100];
	char data[100];
	char src[50];
	char vname[50];
	FILE *f, *g, *h, *m;
	int p, R, s, batch_size, batch, where;

	p = 1000;
	R = 500;
	batch_size = 1;

	strcpy(method,"EMVSS");

	if ( access("/home/cchan40",X_OK) == 0 )
	{
		where = Emory;
		strcpy(master,"/home/cchan40/project/EMSHS");
	}
	else if ( access("/project/qlonglab/changgee",X_OK) == 0 )
	{
		where = HPC;
		strcpy(master,"/home/changgee/project/EMSHS");
	}
	else
	{
		where = LPC;
		strcpy(master,"/home/changgee/project/EMSHS");
	}
	sprintf(home,"%s/EMVS",master);
	sprintf(script,"%s/Sim%d",home,p);
	strcpy(src,"SimEMVSS.R");

	sprintf(data,"%s/datasets",master);

	sprintf(fname,"%s%d",method,p);
	h = fopen(fname,"w");
	chmod(fname,0755);

	sprintf(fname,"%s%dMERGE",method,p);
	m = fopen(fname,"w");
	chmod(fname,0755);

	for ( s=0 ; s<5 ; s++ )
	{
		sprintf(acronym,"%s_%d_%d",method,p,s+1);
		sprintf(vname,"res%s",acronym);

		sprintf(fname,"%s/%s",script,acronym);
		sprintf(line,"%s\n",fname);
		fputs(line,h);

		g = fopen(fname,"w");
		chmod(fname,0755);

		for ( batch=0 ; batch<R ; batch+=batch_size )
		{
			sprintf(fname,"%s/%s_%03d",script,acronym,batch+1);
			if ( where == Emory )
				sprintf(line,"qsub -q fruit.q %s\n",fname);
			else if ( where == HPC )
				sprintf(line,"bsub -q qlonglab -e %s.e -o %s.o < %s\n",fname,fname,fname);
			else
				sprintf(line,"bsub -q cceb_normal -e %s.e -o %s.o < %s\n",fname,fname,fname);
			fputs(line,g);

			f = fopen(fname,"w");
			if ( where == LPC )
				fputs("module load R\n",f);
			else if ( where == HPC )
			{
				fputs("source /etc/profile.d/modules.sh\n",f);
				fputs("module load R-3.3.1\n",f);
			}
			sprintf(Rname,"%s.R",fname);
			sprintf(line,"R --vanilla < %s\n",Rname);
			fputs(line,f);
			fclose(f);

			f = fopen(Rname,"w");
			fputs("library(compiler)\n",f);
			fputs("enableJIT(3)\n",f);
			sprintf(line,"source(\"%s/%s\")\n",home,src);
			fputs(line,f);

			sprintf(line,"r = %d\n",batch_size);
			fputs(line,f);
			sprintf(line,"datapath = \"%s/p%d_%d\"\n",data,p,s+1);
			fputs(line,f);

			if ( p == 1000 )
				fputs("v0 = exp(seq(log(0.0004),log(0.1),length.out=20))\n",f);
			else
				fputs("v0 = exp(seq(log(0.00008),log(0.002),length.out=20))\n",f);
			fputs("v1 = 1000\n",f);
			fputs("eta = 0:1\n",f);

			sprintf(line,"if ( !file.exists(\"%s/%s_%03d\") )\n",script,vname,batch+1);
			fputs(line,f);
			fputs("{\n",f);
			sprintf(line,"  %s = SimEMVSS(r,v0,v1,eta,datapath,batch=%d)\n",vname,batch);
			fputs(line,f);

			sprintf(line,"  save(%s,file=\"%s/%s_%03d\")\n",vname,script,vname,batch+1);
			fputs(line,f);
			fputs("}\n",f);
			fclose(f);
		}
		fclose(g);

		sprintf(fname,"%s/%s_MERGE",script,acronym);
		if ( where == Emory )
			sprintf(line,"qsub -q fruit.q %s\n",fname);
		else if ( where == HPC )
			sprintf(line,"bsub -q qlonglab -e %s.e -o %s.o < %s\n",fname,fname,fname);
		else
			sprintf(line,"bsub -q cceb_normal -e %s.e -o %s.o < %s\n",fname,fname,fname);
		fputs(line,m);

		f = fopen(fname,"w");
		if ( where == LPC )
			fputs("module load R\n",f);
		else if ( where == HPC )
		{
			fputs("source /etc/profile.d/modules.sh\n",f);
			fputs("module load R-3.3.1\n",f);
		}
		sprintf(Rname,"%s.R",fname);
		sprintf(line,"R --vanilla < %s\n",Rname);
		fputs(line,f);
		fclose(f);

		g = fopen(Rname,"w");

		fputs("library(abind)\n",g);
		sprintf(line,"for ( i in 1:%d )\n",R/batch_size);
		fputs(line,g);
		fputs("{\n",g);
		sprintf(line,"  fname = sprintf(\"%s/%s_%%03d\",(i-1)*%d+1)\n",script,vname,batch_size);
		fputs(line,g);
		fputs("  load(fname)\n",g);
		fputs("  if ( i==1 )\n",g);
		fputs("  {\n",g);
		sprintf(line,"    tmp = %s\n",vname);
		fputs(line,g);
		fputs("  }\n",g);
		fputs("  else\n",g);
		fputs("  {\n",g);
		sprintf(line,"    tmp$r = tmp$r + %s$r\n",vname);
		fputs(line,g);
		sprintf(line,"    tmp$FNrate = abind(tmp$FNrate,%s$FNrate)\n",vname);
		fputs(line,g);
		sprintf(line,"    tmp$FPrate = abind(tmp$FPrate,%s$FPrate)\n",vname);
		fputs(line,g);
		sprintf(line,"    tmp$MSTE = abind(tmp$MSTE,%s$MSTE)\n",vname);
		fputs(line,g);
		sprintf(line,"    tmp$MSPE = abind(tmp$MSPE,%s$MSPE)\n",vname);
		fputs(line,g);
		sprintf(line,"    tmp$time = abind(tmp$time,%s$time)\n",vname);
		fputs(line,g);
		fputs("  }\n",g);
		fputs("}\n",g);
		sprintf(line,"%s = tmp\n",vname);
		fputs(line,g);
		sprintf(line,"save(%s,file=\"%s/%s\")\n",vname,home,vname);
		fputs(line,g);

		fclose(g);
	}
	fclose(h);
	fclose(m);
}



